src/main/java/com/aws/greengrass/authorization/WildcardVariableTrie.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.authorization;
+
+import com.aws.greengrass.util.DefaultConcurrentHashMap;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+public class WildcardVariableTrie {
+    protected static final String GLOB_WILDCARD = "*";
+    protected static final String MQTT_WILDCARD = "#";
+    private boolean isTerminal;
+    private boolean isGlobWildcard;
+    private boolean isVariable;
+    private boolean isNull;
+    private boolean matchAll;
+    private final Map<String, WildcardVariableTrie> children =
+            new DefaultConcurrentHashMap<>(WildcardVariableTrie::new);
+
+    /**
+     * Add allowed resources for a particular operation.
+     *
+     * @param subject resource pattern
+     */
+    public void add(String subject) {
+        if (subject == null) {
+            isNull = true;
+            return;
+        }
+        subject = subject.replaceAll("/\\+/", "/*/");
+        add(subject, true);
+    }
+
+    private WildcardVariableTrie add(String subject, boolean isTerminal) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.